### ISSUE TITLE
Basic auth with incorrect credentials doesn't seem to work correctly

### ISSUE THREAD

**Issue by bodgit** [miguelgrinberg/Flask-SocketIO#197]
Hello again,

I'm trying to add basic auth support to my Flask application to authenticate the socket.IO connections and it doesn't seem to work very well in the case of incorrect credentials. I followed the documentation on checking the `current_user` in the connect handler.

Here's the minimal Flask app, it just exposes a `/ping` route that emits a message to connected clients and is itself protected with the same auth, so I know that part is working:

```python
from flask import Flask
from flask.ext.login import LoginManager, UserMixin, current_user, login_required
from flask.ext.socketio import SocketIO

app = Flask(__name__)
login_manager = LoginManager(app)
socketio = SocketIO(app)

class User(UserMixin):

    users = {
        'admin': 'password',
    }

    def __init__(self, username, password):
        self.id = username
        self.password = password

    @classmethod
    def get(cls, id):
        try:
            return User(id, cls.users[id])
        except:
            return None

    def verify_password(self, password):
        return self.password == password


@app.route('/ping')
@login_required
def ping():
    socketio.emit('ping', {})
    return 'ok'


@socketio.on('connect')
def connect_handler():
    if current_user.is_authenticated:
        pass
    else:
        return False


@login_manager.request_loader
def load_user_from_request(request):
    auth = request.authorization
    if auth is not None:
        username, password = auth['username'], auth['password']
        user = User.get(username)
        if not user or not user.verify_password(password):
            return None
        return user
    return None


if __name__ == '__main__':
    socketio.run(app, debug=True)
```

Here's the client:

```python
from socketIO_client import SocketIO, LoggingNamespace, find_callback
import logging

logging.getLogger('requests').setLevel(logging.WARNING)
logging.basicConfig(level=logging.DEBUG)

class Namespace(LoggingNamespace):

    def on_ping(self, *args):
        callback, args = find_callback(args)
        print(args)
        if callback:
            callback(args)

with SocketIO('localhost', 5000, Namespace, auth=('admin', 'password')) as socketIO:
    socketIO.wait()
```

This works fine, but if I set incorrect credentials in the client then it seems to connect but not receive events whereas I would expect it to be disconnected. After the server has tried to send at least one event, if I then Ctrl+C the client, the server logs the following:

```
Traceback (most recent call last):
  File ".../venv-3.5/lib/python3.5/site-packages/eventlet/wsgi.py", line 454, in handle_one_response
    result = self.application(self.environ, start_response)
  File ".../venv-3.5/lib/python3.5/site-packages/flask/app.py", line 1836, in __call__
    return self.wsgi_app(environ, start_response)
  File ".../venv-3.5/lib/python3.5/site-packages/flask_socketio/__init__.py", line 37, in __call__
    start_response)
  File ".../venv-3.5/lib/python3.5/site-packages/engineio/middleware.py", line 47, in __call__
    return self.engineio_app.handle_request(environ, start_response)
  File ".../venv-3.5/lib/python3.5/site-packages/socketio/server.py", line 303, in handle_request
    return self.eio.handle_request(environ, start_response)
  File ".../venv-3.5/lib/python3.5/site-packages/engineio/server.py", line 226, in handle_request
    environ, start_response)
  File ".../venv-3.5/lib/python3.5/site-packages/engineio/socket.py", line 75, in handle_get_request
    start_response)
  File ".../venv-3.5/lib/python3.5/site-packages/engineio/socket.py", line 110, in _upgrade_websocket
    return ws(environ, start_response)
  File ".../venv-3.5/lib/python3.5/site-packages/engineio/async_eventlet.py", line 13, in __call__
    return super(WebSocketWSGI, self).__call__(environ, start_response)
  File ".../venv-3.5/lib/python3.5/site-packages/eventlet/websocket.py", line 127, in __call__
    self.handler(ws)
  File ".../venv-3.5/lib/python3.5/site-packages/engineio/socket.py", line 175, in _websocket_handler
    self.close(wait=False, abort=True)
  File ".../venv-3.5/lib/python3.5/site-packages/engineio/socket.py", line 96, in close
    self.server._trigger_event('disconnect', self.sid)
  File ".../venv-3.5/lib/python3.5/site-packages/engineio/server.py", line 328, in _trigger_event
    return self.handlers[event](*args)
  File ".../venv-3.5/lib/python3.5/site-packages/socketio/server.py", line 438, in _handle_eio_disconnect
    self._handle_disconnect(sid, '/')
  File ".../venv-3.5/lib/python3.5/site-packages/socketio/server.py", line 353, in _handle_disconnect
    if namespace == '/' and self.manager.is_connected(sid, namespace):
  File ".../venv-3.5/lib/python3.5/site-packages/socketio/base_manager.py", line 41, in is_connected
    return sid in self.rooms[namespace][None] and \
KeyError: '/'

127.0.0.1 - - [06/Jan/2016 16:33:05] "GET /socket.io/?EIO=3&sid=a4725ae521664f6fab0f0623051e595f&transport=websocket HTTP/1.1" 500 0 7.935872
```

**Comment by miguelgrinberg** [comment]
Please add logging (pass `logger=True` in your `SocketIO` constructor) to see what the packets look like. When you return `False` from the connect handler, the server sends an ERROR package back to the client. The logging will help confirm that the packet is being sent.

Looks like there is also logging available on the Python client that you are using. It would be useful to see what the exchange of messages looks like from both sides.

**Comment by bodgit** [comment]
The only extra thing I could see logged by adding `logger=True` was the "emitting event..." line shown below:

```
 * Restarting with stat
 * Debugger is active!
 * Debugger pin code: 964-642-917
(5884) wsgi starting up on http://127.0.0.1:5000/
(5884) accepted ('127.0.0.1', 62357)
127.0.0.1 - - [06/Jan/2016 21:09:39] "GET /socket.io/?t=1452114579010-0&EIO=3&transport=polling HTTP/1.1" 200 381 0.002847
(5884) accepted ('127.0.0.1', 62360)
(5884) accepted ('127.0.0.1', 62364)
emitting event "ping" to all [/]
127.0.0.1 - - [06/Jan/2016 21:09:46] "GET /ping HTTP/1.1" 200 117 0.001207
```

The client logged the following:

```
DEBUG:root:localhost:5000/socket.io [transport selected] websocket
DEBUG:root:localhost:5000/socket.io [heartbeat reset]
DEBUG:root:localhost:5000/socket.io [engine.io noop]
DEBUG:root:localhost:5000/socket.io [engine.io pong] b''
...
DEBUG:root:localhost:5000/socket.io [engine.io pong] b''
```

When the client uses the correct credentials, the log looks like this:

```
DEBUG:root:localhost:5000/socket.io [transport selected] websocket
DEBUG:root:localhost:5000/socket.io [heartbeat reset]
DEBUG:root:localhost:5000/socket.io [engine.io noop]
DEBUG:root:localhost:5000/socket.io [engine.io pong] b''
DEBUG:root:localhost:5000/socket.io [engine.io pong] b''
DEBUG:root:localhost:5000/socket.io [engine.io message] b'2["ping",{}]'
DEBUG:root:localhost:5000/socket.io [socket.io packet received] b'2["ping",{}]'
```

**Comment by miguelgrinberg** [comment]
Did you verify that your connect handler returns `False`? Add a print statement to ensure that's happening.

It's odd that the client logging does not include polling. The server logging indicates the client connected with long-polling, so you are either me showing me two different runs that are configured with different transports, or the client-side logging is incomplete and shows a later time, after the long-polling connection was upgraded to websocket.

**Comment by bodgit** [comment]
> Did you verify that your connect handler returns False? Add a print statement to ensure that's happening.

Yes, that's definitely happening. I put a print statement in there and verified it.

> It's odd that the client logging does not include polling. The server logging indicates the client connected with long-polling, so you are either me showing me two different runs that are configured with different transports, or the client-side logging is incomplete and shows a later time, after the long-polling connection was upgraded to websocket.

I don't think I've intentionally missed some logs. I've so far not specified the transport used by the client so I think according to the code by default it is configured with both 'xhr-polling' and 'websocket' transports and picks the best one (try websocket first, then fall back to xhr-polling?). I realised I could turn up the logging in the client so it also shows the HTTP requests. With no transport specified, it looks like this:

```
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): localhost
DEBUG:requests.packages.urllib3.connectionpool:"GET /socket.io/?transport=polling&t=1452117779006-0&EIO=3 HTTP/1.1" 200 119
DEBUG:root:localhost:5000/socket.io [transport selected] websocket
DEBUG:root:localhost:5000/socket.io [heartbeat reset]
DEBUG:root:localhost:5000/socket.io [engine.io noop]
DEBUG:root:localhost:5000/socket.io [engine.io pong] b''
DEBUG:root:localhost:5000/socket.io [engine.io pong] b''
^CDEBUG:root:localhost:5000/socket.io [socket.io disconnect]
```

The matching server log (with the print statement in the connection handler) is:

```
 * Restarting with stat
 * Debugger is active!
 * Debugger pin code: 964-642-917
(8904) wsgi starting up on http://127.0.0.1:5000/
(8904) accepted ('127.0.0.1', 63239)
Not authenticated!
127.0.0.1 - - [06/Jan/2016 22:02:59] "GET /socket.io/?transport=polling&t=1452117779006-0&EIO=3 HTTP/1.1" 200 381 0.000697
(8904) accepted ('127.0.0.1', 63242)
127.0.0.1 - - [06/Jan/2016 22:04:24] "GET /socket.io/?transport=websocket&EIO=3&sid=956db64021464ffd83aeb1c696428fab HTTP/1.1" 200 0 85.012021
```

(The last line takes a while to appear.)

If I force the client to use 'xhr-polling' the client logs look like this:

```
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): localhost
DEBUG:requests.packages.urllib3.connectionpool:"GET /socket.io/?t=1452118304800-0&EIO=3&transport=polling HTTP/1.1" 200 119
DEBUG:root:localhost:5000/socket.io [transport selected] xhr-polling
DEBUG:root:localhost:5000/socket.io [heartbeat reset]
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (2): localhost
DEBUG:requests.packages.urllib3.connectionpool:"POST /socket.io/?t=1452118304815-1&EIO=3&transport=polling&sid=d5d224e8bb1e4718a71fc5865e0271b3 HTTP/1.1" 200 2
DEBUG:requests.packages.urllib3.connectionpool:"GET /socket.io/?t=1452118304817-2&EIO=3&transport=polling&sid=d5d224e8bb1e4718a71fc5865e0271b3 HTTP/1.1" 200 4
DEBUG:root:localhost:5000/socket.io [engine.io pong] b''
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (3): localhost
DEBUG:requests.packages.urllib3.connectionpool:"POST /socket.io/?t=1452118305824-4&EIO=3&transport=polling&sid=d5d224e8bb1e4718a71fc5865e0271b3 HTTP/1.1" 200 2
DEBUG:requests.packages.urllib3.connectionpool:"GET /socket.io/?t=1452118304822-3&EIO=3&transport=polling&sid=d5d224e8bb1e4718a71fc5865e0271b3 HTTP/1.1" 200 4
DEBUG:root:localhost:5000/socket.io [engine.io pong] b''
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (4): localhost
DEBUG:requests.packages.urllib3.connectionpool:"POST /socket.io/?t=1452118306833-6&EIO=3&transport=polling&sid=d5d224e8bb1e4718a71fc5865e0271b3 HTTP/1.1" 200 2
DEBUG:requests.packages.urllib3.connectionpool:"GET /socket.io/?t=1452118305830-5&EIO=3&transport=polling&sid=d5d224e8bb1e4718a71fc5865e0271b3 HTTP/1.1" 200 4
DEBUG:root:localhost:5000/socket.io [engine.io pong] b''
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (5): localhost
DEBUG:requests.packages.urllib3.connectionpool:"POST /socket.io/?t=1452118307842-8&EIO=3&transport=polling&sid=d5d224e8bb1e4718a71fc5865e0271b3 HTTP/1.1" 200 2
DEBUG:requests.packages.urllib3.connectionpool:"GET /socket.io/?t=1452118306839-7&EIO=3&transport=polling&sid=d5d224e8bb1e4718a71fc5865e0271b3 HTTP/1.1" 200 4
DEBUG:root:localhost:5000/socket.io [engine.io pong] b''
^CINFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (6): localhost
DEBUG:requests.packages.urllib3.connectionpool:"POST /socket.io/?t=1452118308279-10&EIO=3&transport=polling&sid=d5d224e8bb1e4718a71fc5865e0271b3 HTTP/1.1" 400 11
DEBUG:root:localhost:5000/socket.io [socket.io disconnect]
```

And the server log is:

```
 * Restarting with stat
 * Debugger is active!
 * Debugger pin code: 964-642-917
(11150) wsgi starting up on http://127.0.0.1:5000/
(11150) accepted ('127.0.0.1', 63509)
Not authenticated!
127.0.0.1 - - [06/Jan/2016 22:11:44] "GET /socket.io/?t=1452118304800-0&EIO=3&transport=polling HTTP/1.1" 200 381 0.002860
127.0.0.1 - - [06/Jan/2016 22:11:44] "POST /socket.io/?t=1452118304815-1&EIO=3&transport=polling&sid=d5d224e8bb1e4718a71fc5865e0271b3 HTTP/1.1" 200 199 0.002406
(11150) accepted ('127.0.0.1', 63512)
127.0.0.1 - - [06/Jan/2016 22:11:44] "GET /socket.io/?t=1452118304817-2&EIO=3&transport=polling&sid=d5d224e8bb1e4718a71fc5865e0271b3 HTTP/1.1" 200 215 0.000257
(11150) accepted ('127.0.0.1', 63515)
127.0.0.1 - - [06/Jan/2016 22:11:45] "POST /socket.io/?t=1452118305824-4&EIO=3&transport=polling&sid=d5d224e8bb1e4718a71fc5865e0271b3 HTTP/1.1" 200 199 0.000290
127.0.0.1 - - [06/Jan/2016 22:11:45] "GET /socket.io/?t=1452118304822-3&EIO=3&transport=polling&sid=d5d224e8bb1e4718a71fc5865e0271b3 HTTP/1.1" 200 215 1.004893
(11150) accepted ('127.0.0.1', 63518)
127.0.0.1 - - [06/Jan/2016 22:11:46] "POST /socket.io/?t=1452118306833-6&EIO=3&transport=polling&sid=d5d224e8bb1e4718a71fc5865e0271b3 HTTP/1.1" 200 199 0.000291
127.0.0.1 - - [06/Jan/2016 22:11:46] "GET /socket.io/?t=1452118305830-5&EIO=3&transport=polling&sid=d5d224e8bb1e4718a71fc5865e0271b3 HTTP/1.1" 200 215 1.005904
(11150) accepted ('127.0.0.1', 63521)
127.0.0.1 - - [06/Jan/2016 22:11:47] "POST /socket.io/?t=1452118307842-8&EIO=3&transport=polling&sid=d5d224e8bb1e4718a71fc5865e0271b3 HTTP/1.1" 200 199 0.000326
127.0.0.1 - - [06/Jan/2016 22:11:47] "GET /socket.io/?t=1452118306839-7&EIO=3&transport=polling&sid=d5d224e8bb1e4718a71fc5865e0271b3 HTTP/1.1" 200 215 1.007462
(11150) accepted ('127.0.0.1', 63524)
127.0.0.1 - - [06/Jan/2016 22:11:48] "POST /socket.io/?t=1452118308279-10&EIO=3&transport=polling&sid=d5d224e8bb1e4718a71fc5865e0271b3 HTTP/1.1" 400 218 0.000305
Traceback (most recent call last):
  File ".../venv-3.5/lib/python3.5/site-packages/eventlet/wsgi.py", line 483, in handle_one_response
    write(b''.join(towrite))
  File ".../venv-3.5/lib/python3.5/site-packages/eventlet/wsgi.py", line 426, in write
    _writelines(towrite)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/socket.py", line 589, in write
    return self._sock.send(b)
  File ".../venv-3.5/lib/python3.5/site-packages/eventlet/greenio/base.py", line 359, in send
    total_sent += fd.send(data[total_sent:], flags)
BrokenPipeError: [Errno 32] Broken pipe

127.0.0.1 - - [06/Jan/2016 22:12:47] "GET /socket.io/?t=1452118307849-9&EIO=3&transport=polling&sid=d5d224e8bb1e4718a71fc5865e0271b3 HTTP/1.1" 400 0 60.005571
```

Does that look saner now?

**Comment by miguelgrinberg** [comment]
Fixed in package python-socketio version 0.9.1. Please upgrade that package to fix this bug.

**Comment by bodgit** [comment]
Yup, that looks good to me, thanks for fixing this.
